[WIP] Add a lexer and parser

dkubb opened the pull request:
This branch adds a ragel based lexer and a racc based parser.

dkubb committed:
Add SQL::Scanner based on ragel

* This is a WIP, just trying to get the right layout and division of
  labour between the lexer and upcoming parser based on racc. This
  will obviously be expanded to include all the tokens in a SELECT
  query, as well as all the tokens in UPDATE, INSERT and DELETE.

mbj commented on lib/sql/scanner.rb.rl:
```diff
@@ -0,0 +1,146 @@
+      @ts = nil
+    end
+
+    def read_chunk
```
I really like this complexity is not anymore within the main `#each`. I think later it could be decoupled from the `Scanner` class altogether.

dkubb committed:
Remove ruby file that will be generated by racc

dkubb committed:
Change ragel to use a scanner instead of a state machine

* This appears to be a better approach since the parsing won't be
  duplicated on the scanner and parser side of things. Ragel is now
  only responsible for tokenizing the input, and Racc will be used
  to parse the tokens.

dkubb committed:
Add SQL::Parser based on racc

* This is a preliminary grammar based on the sql 92 bnf defined at
  http://savage.net.au/SQL/sql-92.bnf.

  Oddly enough it also resembles the racc grammar at:
  https://github.com/omghax/sql/blob/master/lib/sql/parser.racc but
  was not derived from it.

  This scanner/parser combination can now parse simple SQL queries
  like: SELECT * FROM "people" but does not yet do anything
  interesting with the data. Eventually it will create a tree of
  SQL::AST::Node objects.

dkubb committed:
Remove explicit sql/parser require statement

* This should stay removed until the ragel build process can be
  automated on travis.

dkubb committed:
Add subquery support to FROM clauses

dkubb committed:
Add comment about rule naming conventions

dkubb committed:
Change buffer to be flushed after parsing is finished

dkubb committed:
Change SQL::Parser#parse to be a command method

dkubb committed:
Fix rules that do not reduce

dkubb committed:
Remove a qualifier until correlation tracking is supported

dkubb committed:
Add declaration to tell racc what rule to start with

dkubb committed:
Add actions to some parser rules to return ast nodes

dkubb committed:
Add more actions to the parser

* Fix literals to not have quotes in the sexp

dkubb committed:
Add SQL::Parser.parse convenience method

dkubb committed:
Fix parser to only parse the input once

dkubb committed:
Add semicolons to emit statements to be more consistent

* The purpose of the formatting is that this code should work
  with C, Java and Ruby code.

dkubb committed:
Add round-tripping for FROM and AS clauses

dkubb committed:
Refactor scanner to match full unsigned integers

* This allows the parser to focus on matching the whole number, not
  matching each digit and concatenating the strings, and evaluating
  them to get the number.
* Change the identifier emitter in the scanner to remove the leading
  and trailing quotes, as well as any embedded quotes.

dkubb committed:
Add scanner and parser compilation as a spec dependency

dkubb committed:
Add generated files to rubocop exclude list

dkubb committed:
Change version of racc to be compatible with JRuby

dkubb committed:
Move input scanning into a separate method

dkubb committed:
Refactor parser grammar to not require the array index so often

dkubb committed:
Change a SELECT query so the from node is optional

dkubb committed:
Move BNF reference to the top of the file

dkubb committed:
Fix parser to use the correct value

dkubb committed:
Add operator precedence and associativity to the parser

dkubb committed:
Add parenthesis to binary infix operators

* Add comments explaining the algorithm I will used to properly
  parenthesize the operators based on their precedence rules
  compared to the parent, as well as associativity compared to their
  position relative to their siblings.

dkubb committed:
Add mod and pow to the parser

dkubb committed:
Fix failing spec caused by parenthesis changes

* This is only temporary until conditional parenthesis are added.

dkubb committed:
Change specs to use valid SQL queries rather than invalid fragments

* Add DATE operator
* Add String parsing

dkubb committed:
Change ci:metrics to run compilation

dkubb committed:
Fix scanner to emit a token if it is at the end of the string

* Begin adding integration tests for the scanner and parser
  to make sure there are no regressions.

dkubb committed:
Fix scanner to handle multibyte encodings

* This uses the ragel-driven buffer to make sure bytes are aligned
  by character boundaries.

dkubb committed:
Remove extended ascii from the identifier

* The identifier can contain different things for different
  databases. For now match a common subset, and expand later when
  different dialects can be parsed and database specific parsing
  rules can be introduced.

dkubb committed:
Change separator to be more precise in what is matched

dkubb committed:
Change travis to only run specs for now

* The generated code causes the metrics to fail horribly,
  so there need to be a way to turn off scanning of specific files
  that are generated by ragel and racc.

dkubb committed:
Change travis to compile before running spec

dkubb committed:
Add racc 1.4.9 as an explicit dependency for jruby

dkubb committed:
Add rubysl-enumerator as a gem dependency for rbx

dkubb committed:
Fix buffer concatenation to work on ruby 1.9.3

dkubb committed:
Add 2.1.0 to list of rubies that travis expects to fail

dkubb committed:
Upgrade gem dependencies

solnic commented:
Just curious - how far is it from being ready?

mbj commented:
@solnic IMHO it should be ready to do most basic CRUD.

solnic commented:
any update?
